bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/AddonInfoRegistry.java

```diff
@@ -62,14 +65,65 @@ protected void removeAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
     /**
      * Returns the add-on information for the specified add-on ID and locale (language),
      * or {@code null} if no add-on information could be found.
+     * <p>
+     * If more than one provider provides information for the specified add-on ID and locale,
+     * it returns merged information from all such providers.
      *
-     * @param id the ID to be looked for
+     * @param targetId the ID to be looked for
      * @param locale the locale to be used for the add-on information (could be null)
      * @return a localized add-on information object (could be null)
      */
-    public @Nullable AddonInfo getAddonInfo(String id, @Nullable Locale locale) {
-        return addonInfoProviders.stream().map(p -> p.getAddonInfo(id, locale)).filter(Objects::nonNull).findAny()
-                .orElse(null);
+    public @Nullable AddonInfo getAddonInfo(String targetId, @Nullable Locale locale) {
```

Comment:
I think this whole part could be done in a cleaner way. I didn't test, but have a look at this:
```
    public @Nullable AddonInfo getAddonInfo(String targetId, @Nullable Locale locale) {
        return addonInfoProviders.stream().map(p -> p.getAddonInfo(targetId, locale))
                .collect(Collectors.groupingBy(a -> a != null ? a.getUID() : "", Collectors
                        .collectingAndThen(Collectors.reducing((a, b) -> mergeAddonInfo(a, b)), Optional::get)))
                .get(targetId);
    }

    private @Nullable AddonInfo mergeAddonInfo(@Nullable AddonInfo a, @Nullable AddonInfo b) {
        if (a == null) {
            return b;
        } else if (b == null) {
            return a;
        }

        Builder builder = AddonInfo.builder(a);
        if (a.getDescription().isEmpty()) {
            builder.withDescription(b.getDescription());
        }
        if (a.getConnection() == null && b.getConnection() != null) {
            builder.withConnection(b.getConnection());
        }
        Set<String> countries = new HashSet<>(a.getCountries());
        countries.addAll(b.getCountries());
        if (!countries.isEmpty()) {
            builder.withCountries(countries.stream().toList());
        }
        if (a.getConfigDescriptionURI() == null && b.getConfigDescriptionURI() != null) {
            builder.withConfigDescriptionURI(b.getConfigDescriptionURI());
        }
        if (a.getSourceBundle() == null && b.getSourceBundle() != null) {
            builder.withSourceBundle(b.getSourceBundle());
        }
        Set<AddonDiscoveryMethod> discoveryMethods = new HashSet<>(a.getDiscoveryMethods());
        discoveryMethods.addAll(b.getDiscoveryMethods());
        if (!discoveryMethods.isEmpty()) {
            builder.withDiscoveryMethods(discoveryMethods.stream().toList());
        }
        return builder.build();
    }
```

Other fields do not need to be updates because they cannot be null or empty. I also ommitted uid because I think it can be eliminated completely from AddonInfo again, see the previous discussions.

Reply:
> I think this whole part could be done in a cleaner way. I didn't test, but have a look at this:

Does this take into account that there may be not just two providers, but say three or four..?

Reply:
And.. I think perhaps the merge of discoveryMethods might not produce a unique list. ?? I don't recall if I implemented whatever method that `Set.add` uses in order to ensure uniqueness..

Reply:
> Does this take into account that there may be not just two providers, but say three or four..?

Yes, it does.

Reply:
> A HashSet should guarantee each element in the set is unique.

If multiple AddonInfoProviders produce multiple AddonInfos which contain discoveryMethods that are syntactically identical, then nevertheless the base `Object.hash()` will identify those methods as separate objects, so the resulting set would contain duplicate but syntactically identical entries. So I think it is necessary to override the the base `Object.hash()` in order to produce the same hash code based on their discovery syntax rather than their Java instance ids.

Reply:
```
public @Nullable AddonInfo getAddonInfo(String targetId, @Nullable Locale locale) {
        return addonInfoProviders.stream().map(p -> p.getAddonInfo(targetId, locale)).filter(Objects::nonNull)
                .collect(Collectors.groupingBy(a -> a != null ? a.getUID() : "", Collectors
                        .collectingAndThen(Collectors.reducing((a, b) -> mergeAddonInfo(a, b)), Optional::get)))
                .get(targetId);
    }
```

I am guessing that the `a -> a != null ? a.getUID() : ""` can be replaced by `a -> a.getUID()` due to the prior `.filter(Objects::nonNull)` clause. Or ??

Reply:
> If multiple AddonInfoProviders produce multiple AddonInfos which contain discoveryMethods that are syntactically identical, then nevertheless the base `Object.hash()` will identify those methods as separate objects, so the resulting set would contain duplicate but syntactically identical entries. So I think it is necessary to override the base `Object.hash()` in order to produce the same hash code based on their discovery syntax rather than their Java instance ids.

You are correct. I didn't think about it.

Reply:
```
   private @Nullable AddonInfo mergeAddonInfo(@Nullable AddonInfo a, @Nullable AddonInfo b) {
        if (a == null) {
            return b;
        } else if (b == null) {
            return a;
        }
```

And -- due to the prior `.filter(Objects::nonNull)` clause -- probably some of the `@Nullable` and 'if (a/b == null)` checks should be eliminated. ??

Reply:
My Eclipse IDE was complaining about the arguments having to be nullable in the stream. I would agree it shouldn't happen, but kept it to avoid the warnings.

Reply:
> I am guessing that the `a -> a != null ? a.getUID() : ""` can be replaced by `a -> a.getUID()` due to the prior `.filter(Objects::nonNull)` clause. Or ??

My IDE was complaining if I did that.

Reply:
I am playing around with variations on your code, and some give errors in Eclipse, and some in Maven. I didn't yet find one that works error free in both..

Reply:
For me, the code as provided is OK in Eclipse. Trying to compile just the org.openhab.core.addon bundle (`mvn install -pl :org.openhab.core.addon`) gives 2 code analysis error (and 1 warning) for missing pom files, but nothing else, so no related error. I probably should run mvn on the full repository instead of one bundle to make it compile.

Reply:
I just committed a change that compiles in Eclipse and Maven with no warnings.
However nota-bene I have not yet figured out how to test it!

Reply:
`mvn clean install -U -am -DskipTests -pl :org.openhab.core.addon` did the trick, compile with no errors.

Reply:
> I just committed a change that compiles in Eclipse and Maven with no warnings.
> However nota-bene I have not yet figured out how to test it!

I think the only way now would be to create a full distribution. If you then use the REST API addon endpoint to query the addons, you should see connection and countries populated (when the suggestion finder addon has been installed). If you try it without the changes, you will only see connection and countries for installed addons. I don't think discovery methods are propagated to Addon, so will not be visible with the addon endpoint, only with the new suggestions endpoint.

Reply:
I added a JUnit test. And it works. The only minor issue is how to merge the `serviceId` fields (if even required)..

Reply:
By default `serviceId` is `type + "." + id` unless specifically set via the builder. So I would suggest that the merged version should return a specific value if it exists, and the default otherwise. => WDYT?

Reply:
That makes sense. Did you try to find where and under what conditions it is specifically set?

Reply:
> Did you try to find where and under what conditions it is specifically set?

Nope. ;)